pkg/infrastructure/clusterapi/types.go

```diff
@@ -0,0 +1,79 @@
+package clusterapi
+
+import (
+	"context"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/openshift/installer/pkg/asset/installconfig"
+)
+
+// Provider is the base interface that cloud platforms
+// should implement for the CAPI infrastructure provider.
+type Provider interface {
+	// Name provides the name for the cloud platform.
+	Name() string
+}
+
+// PreProvider defines the PreProvision hook, which is called prior to
+// CAPI infrastructure provisioning.
+type PreProvider interface {
+	// PreProvision is called before provisioning using CAPI controllers has begun
+	// and should be used to create dependencies needed for CAPI provisioning,
+	// such as IAM roles or policies.
+	PreProvision(ctx context.Context, in PreProvisionInput) error
+}
+
+// PreProvisionInput collects the args passed to the PreProvision call.
+type PreProvisionInput struct {
+	InfraID       string
+	InstallConfig *installconfig.InstallConfig
+}
+
+// IgnitionProvider handles preconditions for bootstrap ignition and
+// generates ignition data for the CAPI bootstrap ignition secret.
+//
+// WARNING! Low-level primitive. Use only if absolutely necessary.
+type IgnitionProvider interface {
+	Ignition(ctx context.Context, in IgnitionInput) (string, error)
+}
+
+// IgnitionInput collects the args passed to the IgnitionProvider call.
+type IgnitionInput struct {
+	BootstrapIgnData string
```

Comment:
I'm curious: why string? I understand that JSON spec guarantees it can be encoded UTF-8, but I wonder why casting back and forth.

Reply:
Fixed! I was just coalescing the types based on the signature values of the `InjectInstallInfo` func. I updated that func so it returns []byte and eliminated the casting. Thanks!